test/integration/client/client_test.go

```diff
@@ -802,3 +807,80 @@ func TestSelfLinkOnNamespace(t *testing.T) {
 
 	runSelfLinkTestOnNamespace(t, c, "default")
 }
+
+func TestApplyWithApplyConfiguration(t *testing.T) {
+	deployment := appsv1ac.Deployment("nginx-deployment-3", "default").
+		WithSpec(appsv1ac.DeploymentSpec().
+			WithSelector(metav1ac.LabelSelector().
+				WithMatchLabels(map[string]string{"app": "nginx"}),
+			).
+			WithTemplate(corev1ac.PodTemplateSpec().
+				WithLabels(map[string]string{"app": "nginx"}).
+				WithSpec(corev1ac.PodSpec().
+					WithContainers(
+						corev1ac.Container().
+							WithName("nginx").
+							WithImage("nginx:1.14.2").
+							WithStdin(true).
+							WithPorts(
+								corev1ac.ContainerPort().
+									WithContainerPort(8080).
+									WithProtocol(v1.ProtocolTCP),
+							).
+							WithResources(corev1ac.ResourceRequirements().
+								WithLimits(v1.ResourceList{
+									v1.ResourceCPU:    resource.MustParse("4"),
+									v1.ResourceMemory: resource.MustParse("32Gi"),
+								}),
+							),
+					),
+				),
+			),
+		)
+	testServer := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins", "ServiceAccount"}, framework.SharedEtcd())
+	defer testServer.TearDownFn()
+
+	c := clientset.NewForConfigOrDie(testServer.ClientConfig)
+
+	// Test apply to spec
+	obj, err := c.AppsV1().Deployments("default").Apply(context.TODO(), deployment, metav1.ApplyOptions{FieldManager: "test-mgr", Force: true})
+	if err != nil {
+		t.Fatalf("unexpected error when applying manifest for Deployment: %v", err)
+	}
+	if obj.Spec.Template.Spec.Containers[0].Image != "nginx:1.14.2" {
+		t.Errorf("expected image %s but got %s", "nginx:1.14.2", obj.Spec.Template.Spec.Containers[0].Image)
+	}
+	cpu := obj.Spec.Template.Spec.Containers[0].Resources.Limits[v1.ResourceCPU]
+	if cpu.Value() != int64(4) {
+		t.Errorf("expected resourceCPU limit %d but got %d", 4, cpu.Value())
+	}
+
+	// Test apply to status
+	statusApply := appsv1ac.Deployment("nginx-deployment-3", "default").
+		WithStatus(appsv1ac.DeploymentStatus().
+			WithConditions(
+				appsv1ac.DeploymentCondition().
+					WithType(appsv1.DeploymentReplicaFailure).
+					WithStatus(v1.ConditionUnknown).
+					WithLastTransitionTime(metav1.Now()).
+					WithLastUpdateTime(metav1.Now()).
+					WithMessage("apply status test").
+					WithReason("TestApplyWithApplyConfiguration"),
+			),
```

Comment:
Yeah this is going to be great for controllers. Appending was clearly the right thing to do--imagine passing the status builder around to different parts of the controller. (just thinking out loud)

What happens if I e.g. add the same condition twice? Will the server give an error? It seems like a plausible mistake.

Reply:
Agreed. I think it's good we arrived at a builder pattern that is more clearly distinct from just "setters that return the receiver" since it opens up some additional usage patterns that can complement what would already be possible by directly manipulating the struct. It will be interesting to see what developers do with it in controllers for more complex cases.

> What happens if I e.g. add the same condition twice? Will the server give an error? It seems like a plausible mistake.

Oh, so is kinda interesting to think about. The server does return a validation error. It's the same error a developer would receive if they passed a struct with the same condition twice:

```
apiserver received an error that is not an metav1.Status: &errors.errorString{s:"failed to create typed patch object: .status.conditions: duplicate entries for key [type=\"ReplicaFailure\"]"}: failed to create typed patch object: .status.conditions: duplicate entries for key [type="ReplicaFailure"]
```


So this is roughly on part with go structs.  But maybe it could be better? I'm just thinking about this now, but it would be possible to generate the builders to be aware of the keys of associative lists and do something with that knowledge. I haven't thought enough about the implications of pushing that type of knowledge to the client-side, and if that's a good idea or not. If we did something like that it might be that we introduce "Put<Field>" builders? I don't know if we'd want to deprecate the "With<Field>" builder for that field or not then..

Reply:
Yeah, we can possibly do that now that we're embedding the schema in the clients :-) Wouldn't have been an option without it.

Reply:
I think the server message is clear enough and a client error wouldn't be better unless we could make it a compile time error, which I don't think we can?

Reply:
Yeah, server error is fine, I agree. I was referring to the ability to overwrite the entry (which we can do if we have schema knowledge)

Reply:
oh, that's an interesting idea. But I think it might cover up a client bug that they would prefer to know about?

Reply:
Yeah, I agree. The whole conflict thing is to help people realize that they overwrite values by accident.
On the other hand, it's a little challenging to find/remove/replace if that's really what you want to do. We could help them find possible duplicate entries, and/or remove by key in associative lists.

Reply:
I'll added this to the list of things to discuss more post GA